### PR TITLE
Add some retry to static IP create, I've seen it flake recently

### PR DESCRIPTION
**Pull request by brendandburns**
We've seen ip creation flake at least twice in recent e2e runs, when in doubt, add some retry.

@quinton-hoole @ixdy @zmerlynn

### CODE REVIEW

File: cluster/gce/util.sh [hunk 1]
```diff
@@ -652,7 +672,8 @@ function kube-up {
   # so extract the region name, which is the same as the zone but with the final
   # dash and characters trailing the dash removed.
   local REGION=${ZONE%-*}
-  MASTER_RESERVED_IP=$(gcloud compute addresses create "${MASTER_NAME}-ip" \
+  create-static-ip "${MASTER_NAME-ip}"
```
Comment (ixdy) [inline cluster/gce/util.sh]: this should be `${MASTER_NAME}-ip` (-ip outside braces)

File: cluster/gce/util.sh [hunk 2]
```diff
@@ -331,6 +331,26 @@ function wait-for-jobs {
...
+  while true; do
+    if ! gcloud compute addresses create "$1" \
+      --project "${PROJECT}" \
+      --region "${REGION}" -q --format yaml; then
+      if (( attempt > 5 )); then
+        echo -e "${color_red}Failed to create static ip $1 ${color_norm}"
+        exit 2
+      fi
+     attempt=$(($attempt+1)) 
+     echo -e "${color_yellow}Attempt $attempt failed to create static ip $1. Retrying.${color_norm}"
```
Comment (ixdy) [inline cluster/gce/util.sh]: echo >&2 to mirror fix in #15209

File: cluster/gce/util.sh [hunk 3]
```diff
@@ -331,6 +331,26 @@ function wait-for-jobs {
...
 
+# Robustly try to create a static ip.
+# $1: The name of the ip to create
+function create-static-ip {
+  detect-project
+  local attempt=0
+  while true; do
+    if ! gcloud compute addresses create "$1" \
+      --project "${PROJECT}" \
+      --region "${REGION}" -q --format yaml; then
```
Comment (ixdy) [inline cluster/gce/util.sh]: probably don't want --format yaml here

File: cluster/gce/util.sh (hunk 1 above)
Comment (brendandburns) [inline cluster/gce/util.sh]: good catch.  done.

File: cluster/gce/util.sh (hunk 2 above)
Comment (brendandburns) [inline cluster/gce/util.sh]: done.

File: cluster/gce/util.sh (hunk 3 above)
Comment (brendandburns) [inline cluster/gce/util.sh]: done. (and piped to /dev/null)

File: pkg/apis/experimental/v1alpha1/types_swagger_doc_generated.go [hunk 4]
```diff
@@ -356,7 +356,8 @@ func (JobStatus) SwaggerDoc() map[string]string {
 }
 
 var map_NodeUtilization = map[string]string{
-	"": "NodeUtilization describes what percentage of a particular resource is used on a node.",
+	"":      "NodeUtilization describes what percentage of a particular resource is used on a node.",
```
Comment (quinton-hoole) [inline pkg/apis/experimental/v1alpha1/types_swagger_doc_generated.go]: This change looks unrelated.  Rebase?

File: cluster/gce/util.sh [hunk 5]
```diff
@@ -331,6 +331,26 @@ function wait-for-jobs {
   fi
 }
 
+# Robustly try to create a static ip.
+# $1: The name of the ip to create
+function create-static-ip {
```
Comment (quinton-hoole) [inline cluster/gce/util.sh]: nit: This is going to retry in rapid succession in a tight loop.  Please insert an exponential backoff (even as simple as a "sleep ${attempt * 5}".

**Comment by quinton-hoole** [comment]
Thanks @brendandburns !  Minor nit then LGTM.

File: pkg/apis/experimental/v1alpha1/types_swagger_doc_generated.go (hunk 4 above)
Comment (brendandburns) [inline pkg/apis/experimental/v1alpha1/types_swagger_doc_generated.go]: ugh, this is fixed by https://github.com/kubernetes/kubernetes/pull/15242

**Comment by brendandburns** [comment]
Rebased.

File: cluster/gce/util.sh (hunk 5 above)
Comment (brendandburns) [inline cluster/gce/util.sh]: done (and added to the other retry loop too)

**Comment by brendandburns** [comment]
nit addressed.  ptal.

thanks!

File: cluster/gce/util.sh [hunk 6]
```diff
@@ -331,6 +331,27 @@ function wait-for-jobs {
...
+  detect-project
+  local attempt=0
+  while true; do
+    if ! gcloud compute addresses create "$1" \
```
Comment (ixdy) [inline cluster/gce/util.sh]: this is missing a `fi`.

`/jenkins-master-data/jobs/kubernetes-pull-build-test-e2e-gce/workspace@7/hack/e2e-iternal/../../cluster/gce/util.sh: line 352: syntax error near unexpected token `done'`

File: cluster/gce/util.sh [hunk 7]
```diff
@@ -331,6 +331,27 @@ function wait-for-jobs {
...
+      --region "${REGION}" -q > /dev/null; then
+      if (( attempt > 5 )); then
+        echo -e "${color_red}Failed to create static ip $1 ${color_norm}"
+        exit 2
+      fi
+      attempt=$(($attempt+1)) 
+      echo -e "${color_yellow}Attempt $attempt failed to create static ip $1. Retrying.${color_norm}" >&2
+      sleep $(($attempt * 5))
+    else
+     break
```
Comment (ixdy) [inline cluster/gce/util.sh]: minor nit (since we need a `fi` too): indent this `break` an additional space

File: cluster/gce/util.sh (hunk 6 above)
Comment (brendandburns) [inline cluster/gce/util.sh]: done (oops)

File: cluster/gce/util.sh (hunk 7 above)
Comment (brendandburns) [inline cluster/gce/util.sh]: done.

**Comment by brendandburns** [comment]
Comments addressed, please re-check.

Thanks!
--brendan

File: cluster/gce/util.sh [hunk 8]
```diff
@@ -331,6 +331,28 @@ function wait-for-jobs {
...
 
+# Robustly try to create a static ip.
+# $1: The name of the ip to create
+function create-static-ip {
+  detect-project
+  local attempt=0
+  while true; do
+    if ! gcloud compute addresses create "$1" \
+      --project "${PROJECT}" \
+      --region "${REGION}" -q > /dev/null; then
```
Comment (ixdy) [inline cluster/gce/util.sh]: just noticed that `REGION` is declared local on L677, so this will also probably fail

Comment (ixdy) [inline cluster/gce/util.sh]: except that it worked? maybe I don't understand how local works.
in any case, having a `local REGION=${ZONE%-*}` in this function might be a good idea too.

Comment (brendandburns) [inline cluster/gce/util.sh]: added as a second parameter to `create-static-ip`

**Comment by roberthbailey** [comment]
@quinton-hoole - want to lgtm?

**Comment by thockin** [comment]
@a-robinson does this still apply?

File: cluster/gce/util.sh [hunk 9]
```diff
@@ -347,12 +371,13 @@ function create-firewall-rule {
...
-          echo -e "${color_red}Failed to create firewall rule $1 ${color_norm}" >&2
-          exit 2
-        fi
-        echo -e "${color_yellow}Attempt $(($attempt+1)) failed to create firewall rule $1. Retrying.${color_norm}" >&2
-        attempt=$(($attempt+1))
+      if (( attempt > 5 )); then
+        echo -e "${color_red}Failed to create firewall rule $1 ${color_norm}" >&2
+        exit 2
+      fi
+      echo -e "${color_yellow}Attempt $(($attempt+1)) failed to create firewall rule $1. Retrying.${color_norm}"
```
Comment (ixdy) [inline cluster/gce/util.sh]: >&2

File: cluster/gce/util.sh [hunk 10]
```diff
@@ -333,6 +333,30 @@ function wait-for-jobs {
...
+# $2: The name of the region to create the ip in.
+function create-static-ip {
+  detect-project
+  local attempt=0
+  local REGION="$2"
+  while true; do
+    if ! gcloud compute addresses create "$1" \
+      --project "${PROJECT}" \
+      --region "${REGION}" -q > /dev/null; then
+      if (( attempt > 5 )); then
```
Comment (ixdy) [inline cluster/gce/util.sh]: probably doesn't matter too much, but this'll actually try 6 times

File: cluster/gce/util.sh [hunk 11]
```diff
@@ -333,6 +333,30 @@ function wait-for-jobs {
...
+        echo -e "${color_red}Failed to create static ip $1 ${color_norm}"
```
Comment (ixdy) [inline cluster/gce/util.sh]: >&2

**Comment by a-robinson** [comment]
@thockin - Yes, this still applies. It's worth it to let the master have a static IP address at all times.

File: cluster/gce/util.sh (hunk 9 above)
Comment (brendandburns) [inline cluster/gce/util.sh]: done

File: cluster/gce/util.sh (hunk 10 above)
Comment (brendandburns) [inline cluster/gce/util.sh]: done.

File: cluster/gce/util.sh (hunk 11 above)
Comment (brendandburns) [inline cluster/gce/util.sh]: dine.

**Comment by brendandburns** [comment]
Comments addressed.  ptal.

thanks!
--brendan

File: test/e2e/google_compute.go [hunk 12]
```diff
@@ -32,15 +33,22 @@ func createGCEStaticIP(name string) (string, error) {
...
+	for retries := 0; retries < 5; retries++ {
+		output, err = exec.Command("gcloud", "compute", "addresses", "create",
+			name, "--project", testContext.CloudConfig.ProjectID,
+			"--region", "us-central1", "-q").CombinedOutput()
+		if err == nil {
+			break
+		}
 		glog.Errorf("Creating static IP with name:%s in project: %s", name, testContext.CloudConfig.ProjectID)
 		glog.Errorf("output: %s", output)
+		time.Sleep(2 * time.Second)
```
Comment (quinton-hoole) [inline test/e2e/google_compute.go]: nit:Similar comment to previous one.  Please use exponential back along the lines of "sleep $((retries * 5))" rather than an arbitrary static sleep period.

nit: Why 'retries' here and 'attempt' above.  Prefer to be consistent.

Comment (brendandburns) [inline test/e2e/google_compute.go]: done & done.

**Comment by brendandburns** [comment]
comment addressed, please take another look.

thanks!
--brendan

**Comment by quinton-hoole** [comment]
LGTM!

**Comment by wojtek-t** [comment]
Jenkins failure was issue with starting cluster

@k8s-bot test this please

**Comment by wojtek-t** [comment]
e2e test failure is flake:
"Job should run a job to completion when tasks sometimes fail and are locally restarted"

@k8s-bot test this please
